lib/sprockets/utils.rb

```diff
@@ -97,11 +96,21 @@ def string_end_with_semicolon?(str)
     #
     # Returns buf String.
     def concat_javascript_sources(buf, source)
-      if buf.bytesize > 0
-        buf << ";" unless string_end_with_semicolon?(buf)
-        buf << "\n" unless buf.end_with?("\n")
+      if source.bytesize > 0
+        buf << source
+
+        # If the source contains non-ASCII characters, indexing on it becomes O(N).
+        # This will lead to O(N^2) performance in string_end_with_semicolon?, so we should use 32 bit encoding to make sure indexing stays O(1)
+        source = source.encode(Encoding::UTF_32LE) unless source.ascii_only?
+
+        if !string_end_with_semicolon?(source)
+          buf << ";\n"
```

Comment:
Self answer: https://github.com/rails/sprockets/commit/4d00bdfd54ad47c6bb90859e5fe7a51c495121d6 it seems to be to prevent shifting lines in source maps.

Maybe it would be simple to always add that line, and always shift the source maps by 1.